src/core/thread/discover_scanner.cpp

```diff
@@ -152,6 +152,18 @@ Error DiscoverScanner::Discover(const Mac::ChannelMask &aScanChannels,
     mScanChannel = Mac::ChannelMask::kChannelIteratorFirst;
     mState       = (mScanChannels.GetNextChannel(mScanChannel) == kErrorNone) ? kStateScanning : kStateScanDone;
 
+    // For rx-off-when-idle device, temporary enable receiver during discovery procedure.
+    if (!Get<Mle>().IsRxOnWhenIdle())
+    {
+#if OPENTHREAD_CONFIG_MAC_CSL_RECEIVER_ENABLE
+        if ((mPrevCslPeriod = Get<Mac::Mac>().GetCslPeriod()))
+        {
+            Get<Mac::Mac>().SetCslPeriod(0);
+        }
```

Comment:
I think this model can have side-effects. For example, what if the CSL period happen to be changed in middle of discover scan (user can call `otLinkCslSetPeriod()` at anytime). We save the value here and then re-apply the old value at the end.

-----

> turning off CSL if the period is not equal to 0 during discovery (to not include CSL IE and Auxilary Security Header in the MLE Discovery Request)

I think it may be better to find a different way to achieve this (without `SetCslPeriod(0)`)?

- Maybe can change `Mac::IsCslEnabled(void)` to also check if `Mac` is not `RxOnWhenIdl` (it now only checks with `Mle`). If not CSL cpable then no CSL IE should be appended.

- Or another way, can update `MeshForwarder::CalcIePresent()` to check if `Message` is discover msg using sub-type `kSubTypeMleDiscoverRequest` and return `false` (so no IE is appended).

Reply:
Make sense, will give it a try.